### ISSUE TITLE
Mutex#lock and Mutex#synchronize are not interruptible by Thread#kill

### ISSUE THREAD

**Issue by eregon** [jruby/jruby#5476]
### Environment

* JRuby version: `jruby 9.2.5.0-SNAPSHOT (2.5.0) 2018-11-27 ebbb750 OpenJDK 64-Bit Server VM 25.171-b10 on 1.8.0_171-b10 +jit [linux-x86_64]`

Run:

```
ruby -e 'm=Mutex.new; m.lock; t=Thread.new{m.lock}; sleep 1; t.kill; t.join; p :done' 
```

### Expected Behavior

```
:done
```

### Actual Behavior

It hangs on JRuby.
3 Mutex specs have been tagged in https://github.com/jruby/jruby/pull/5475 to avoid hanging.

**Comment by headius** [comment]
Huh. Must have regressed...I'm sure this worked before.

**Comment by andy-twosticks** [comment]
Hangs for me on 9.1.8.0, 9.1.17.0, 9.2.0.0 and 9.2.1.0-dev.

I'd rather like this to work on 9.1.17.0, since that's the "stable" release.  I _think_ it might be tangentally responsible for a problem I have, where threads are being held onto between rspec sessions despite being supposedly killed at the end of the test?

**Comment by eregon** [comment]
Couldn't the semantics just be try to lock in an interruptible manner, and if it is interrupted, check for `Thread#raise`, otherwise just try to lock (checking interrupts) again? That's essentially what we do in TruffleRuby.

**Comment by headius** [comment]
@andy-twosticks: Do you have a real-world case freezing due to this? I am leaning toward going back to interruptible with some status check. As @eregon mentions, checking for an interrupt event may be sufficient, but I'm not sure I know all the cases that are expected to wake a lock-blocked thread.

**Comment by headius** [comment]
@eregon The main concern I have with that approach is all the other JVM interrupt cases that aren't related to Ruby. Since we're a JVM language, we have to make decisions about whether to ignore non-Ruby interrupts or allow them to cancel the blocking operation being performed. There's no perfect answer.

We have had multiple user reports in the past where blocking operations were or were not interruptible via standard JVM mechanisms. Often we have to make a judgment call which way to go, but simply forcing Ruby semantics on users is not usually the first choice.

**Comment by headius** [comment]
Seems like we would need to either support Java interrupt semantics or Ruby interrupt semantics. It would be possible for us to extend Thread and override `interrupt` to set a Ruby thread event, so Java interrupts would properly interrupt a lock, but this basically puts us back in the boat we were trying to fix in #4261 (where an interrupt intended to interrupt regex processing woke up a lock instead).

So summarizing options:

* No interrupt: prevents threads waiting on a mutex from being interrupted in the usual Ruby ways.
* Allow interrupt, no status check: allows normal JVM interrupting to work but risks spurious wakeups from other JVM events and signals.
* Allow interrupt, check for Ruby event: prevents normal JVM interrupting from working unless a Ruby thread event has been set. Ruby interrupts work as expected in this issue's description.

**Comment by andy-twosticks** [comment]
@headius -- I don't have a real-world freeze (although of course I have no way of knowing if any of the libraries I am using are trying to kill a thread that's held by a Mutex...).  Then again, the code that freezes isn't live yet.  

All I have is an rspec test that causes the test suite to freeze _sometimes_ (or would, if I hadn't added a timeout to the test) and other times just crashes from some other part of the suite (and the trace says the crash happened inside a Mutex).

My current working theory is that somehow repeated runs of the test suite build up dead threads until I hit some sort of limit and I either get a hang or a crash.  I know that sounds wild, but it's the best I have right now.

Also, when I kill my threads from the test (which is when it hangs, when it hangs) my code should not be inside a Mutex.  So I'm not sure if I'm experiencing this bug or a different one.

I'm currently trying to duplicate this outside of rspec.  If I come up with anything concrete I'll raise a seperate issue, but please don't hold your breath; it's proving highly problematic, and I'm going to have to switch to another project any time now.

**Comment by headius** [comment]
@andy-twosticks The fix isn't difficult...it basically restores the original logic that does a thread poll, but with the additional wrinkle that if the thread was not interrupted for a Ruby threading reason it will resume attempting to acquire the lock. I'm on the fence about that, because it doesn't seem like we can make both Java interrupters and Ruby interrupters happy.

```diff
diff --git a/core/src/main/java/org/jruby/ext/thread/Mutex.java b/core/src/main/java/org/jruby/ext/thread/Mutex.java
index 196390653d..e3d55f23a4 100644
--- a/core/src/main/java/org/jruby/ext/thread/Mutex.java
+++ b/core/src/main/java/org/jruby/ext/thread/Mutex.java
@@ -97,7 +97,18 @@ public class Mutex extends RubyObject implements DataType {
             // failed to acquire, proceed to sleep and block
             try {
                 thread.enterSleep();
-                thread.lock(lock);
+
+                // Loop until lock is acquired or we are interrupted with a thread event.
+                while (true) {
+                    try {
+                        thread.lockInterruptibly(lock);
+                        break;
+                    } catch (InterruptedException ie) {
+                        // Check thread events; if one is present, it will propagate. Otherwise interrupt was spurious.
+                        // See jruby/jruby#5476
+                        thread.pollThreadEvents(context);
+                    }
+                }
             } finally {
                 thread.exitSleep();
             }
```

**Comment by headius** [comment]
I have filed #5682 which applies the above patch.

At this point I'm leaning back toward making all locks be interruptible by both Java and Ruby. The original accommodation for LogStash was in order to allow them to interrupt threads that appeared hung (based on some heuristic) without also interrupting threads validly blocked on a lock. If the contract of a Mutex in Ruby were that it's not interruptible from either Ruby or C, the current logic makes sense, but my reading of the world says that both Ruby and C can interrupt a thread blocking on a contended Mutex, which is how it should work in Java.

You out there @andrewvc? Or perhaps @colinsurprenant knows who else could discuss the original issue #4261?

**Comment by headius** [comment]
I'm going ahead with the changes in #5683, which align Mutex and ConditionVariable much more closely with CRuby behavior.

**Comment by colinsurprenant** [comment]
@headius just came across this issue, for some reason it evaded my radar. This looks reasonable, I am actually following up on some interruption issues, will let you know if we have concerns related to this.

**Comment by headius** [comment]
@colinsurprenant The fix for this issue triggered a regression that led to #6326 and #6405. I've pushed a fix in #6407, have a look!
